File: python_coding/3.BFS&DFS/BOJ_2606.py
# 바이러스

# 연결된 걸 기준으로 모든 부분을 체크 해봐야함
# dfs가 적합하겠다고 생각

# dfs
def dfs(graph,v,visited):
    visited[v] = True
    for i in graph[v]:
        if not visited[i]:
            dfs(graph,i,visited)

n = int(input())
l = int(input())


# 인접 리스트 형태로 변경
com = [[]*n for _ in range(n+1)]
for i in range (l):
    a,b = map(int,input().split())
    com[a].append(b)
    com[b].append(a)


visited = [False]*(n+1)
dfs(com,1,visited)
print(visited.count(True)-1)

File: java_coding/data_structure/bfs_dfs/BOJ_2606.java
package java_coding.data_structure.bfs_dfs;

import java.io.*;
import java.util.*;

public class BOJ_2606 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        int n = scanner.nextInt(); 
        int m = scanner.nextInt(); 

        ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < m; i++) {
            int u = scanner.nextInt();
            int v = scanner.nextInt();
            graph.get(u).add(v);
            graph.get(v).add(u);
        }

        boolean[] visited = new boolean[n + 1];
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(1); 
        visited[1] = true;
        int count = 0;

        while (!queue.isEmpty()) {
            int current = queue.poll();
            count++;

            for (int next : graph.get(current)) {
                if (!visited[next]) {
                    queue.offer(next);
                    visited[next] = true;
                }
            }
        }

        System.out.println(count - 1);
    }
}